Upstream fixes

thomaslitton opened the pull request:
## Pull Request Checklist

**Is this in reference to an existing issue?**

#### General

- [ ] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [x] Tests

- [x] Add the plugin to the README

- [x] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose

#### Known Compatibility Issues

thomaslitton committed:
Added a check that pulls the last value for any azure metric and compares it against thresholds.

thomaslitton committed:
Separated the resource id and resource name options

thomaslitton committed:
Add MSI options.

Added options --use-assigned-identity and --local-auth-port to use
the VMs assigned identity (see MSI docs/portal/cli) and change
the default port if necessary.

thomaslitton committed:
Fixed an issue where the date calculation
was not correct for non-utc timezones.

thomaslitton committed:
Fixed an issue where the date calculation
was not correct for non-utc timezones.

thomaslitton committed:
2 quick changes:

1. bumped the version
2. Removed unused function.

thomaslitton committed:
Fixing rubocop issue

thomaslitton committed:
Fixed a duplicate short parameter (client and critical)
Fixed the version
Added the change log entry.
Added documentation to the readme.

majormoses commented:
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

majormoses commented on CHANGELOG.md:
```diff
@@ -5,6 +5,9 @@ This CHANGELOG follows the format located [here](https://github.com/sensu-plugin
 
 ## [Unreleased]
 
+## [2.1.0] - 2018-03-16
```
per our [changelog conventions](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md) maintainers version and date the releases post acceptance. There are several reasons for this:
- PR review and merge is not FIFO
- there is no SLA on on times waiting for review, I try to review within 24-48 hours. While I often review much quicker than that there are times where it takes me several days to get to them
- maintainer might disagree on the impact of a particular change
- As PRs do not necessarily get merged in the order they receive not including the version reduces the number of times we have to ask for a rebase or updating the value.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+         proc: proc { |val| val.to_i }
+
+  def run
+    if config[:resource_id].to_s.empty? && config[:resource_name].to_s.empty?
```
based on the error message it should be `||` not `&&` update the logic or message to indicate what is right.

majormoses replied:
Given bellow this implies the logic should change

thomaslitton replied:
If it were an or, that would require the user to provide both correct?  Only one is needed.

majormoses replied:
ah I was thrown off by the verbiage in the message. I will comment below on what I think we can do to make it clearer.

majormoses commented on CHANGELOG.md:
```diff
@@ -5,6 +5,9 @@ This CHANGELOG follows the format located [here](https://github.com/sensu-plugin
 ## [Unreleased]
 
+## [2.1.0] - 2018-03-16
+- added monitoring metric check
```
Ideally a little more info will help people a monitoring check is quite ambiguous. Maybe something like this:
```
- check-azurerm-monitor-metric.rb: allows you to check against azure metric thresholds for a particular resource id or name (@thomaslitton)
```

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+         description: 'Port used to authenticate when using the local identity via Managed Service Identity (MSI)',
+         short: '-o PORT',
+         long: '--local-auth-port PORT',
+         default: '50342'
```
why is this a string? It probably should be an integer right?

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+          critical "Metric #{metric_val[:metric_name]} is #{metric_val[:value]}"
+        elsif config[:warning_over] && metric_val[:value] > config[:warning_over].to_f
+          warning "Metric #{metric_val[:metric_name]} is #{metric_val[:value]}"
+        elsif config[:critical_under] && metric_val[:value] < config[:critical_under].to_f
```
So you could not say specify a utilization range then? I think there are use cases where you want things to stay in a range. Queued worker messages for backend batch or async processing work comes to mind. Network utilization from a NAT comes to mind, if there is ever under a certain amount of traffic there is something horribly wrong if its too high you need to look into scale your NAT throughput.

I am not asking you to add it just talking through some things that someone might want in the future. You might find you have use cases you did not think about and want this even if its in another PR.

If we wanted to do this we could just keep track of the critical, warnings for each and at the end we could include all crit and/or warning

thomaslitton replied:
I'll try to fit this change in.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+        ts[:data].reverse_each do |metric_value|
+          if metric_value[config[:aggregation].to_sym]
+            values << {
+              value: metric_value[config[:aggregation].to_sym].to_f,
```
why cast it to a symbol and then a float?

thomaslitton replied:
It is not.  It is casting the option used as the key to a symbol, and the value from the hash to a float.

majormoses replied:
you are right, this is what happens when I spend too much time reviewing PRs.

majormoses replied:
you could do a conditional assignment above to reduce code duplication but by no means a requirement.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+
+  def metric_response
+    auth_header = if config[:use_assigned_identity]
+                    uri = URI.parse("http://localhost:#{config[:local_auth_port]}/oauth2/token?resource=https://management.azure.com/")
```
It's been my experience with azure that they have a bunch of different portals and they change them regularly perhaps we should expose this so if they ever change it people can update their runtime options and unblock themselves without waiting for a fixed release. Not a requirement but just a thought.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+                  end
+
+    begin
+      url = "https://management.azure.com#{resource}/providers/microsoft.insights/metrics?" \
```
seeing as we have seen this url twice in the file we should at least variablize it if not making it a cli arg.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,305 @@
+
+  def timespan
+    # 10 min.  This should be enough time to capture the last value, without wasting API credits
+    start_date = Time.now.utc - 600
```
I think that the timespan should be configurable, I can't say I know much about how azure does metrics but I do look at cloudwatch metrics in aws in shorter intervals than this for some metrics.

majormoses commented on lib/sensu-plugins-azurerm/version.rb:
```diff
@@ -1,7 +1,7 @@
   module Version
     MAJOR = 2
-    MINOR = 0
+    MINOR = 1
```
see previous comment about maintainers bumping versions post acceptance.

majormoses reviewed (changes requested):
This looks really great, I have some small suggestions to make. I really appreciate the tests.

thomaslitton committed:
Fixes from review:

1.  Reverted version update.
2.  Added option for API base url
3.  Added option for resource url when authenticating locally.
4.  Added option for lookback timespan.
5.  Make the port a number.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -150,6 +157,20 @@ class CheckAzurermMonitorMetric < Sensu::Plugin::Check::CLI
+
   def run
     if config[:resource_id].to_s.empty? && config[:resource_name].to_s.empty?
       unknown 'resource id or resource name/group/type/namespece and subscription id must be provided'
```
Given the above I think we can improve this, something along the lines of:
```
unknown "resource id, resource name/group/type/namespace, and subscription id are required. You passed: resource_id: #{config[:resource_id]}, resource_name: #{config[:resource_name]}"
```

Also the error message indicates that subscription id must be specified but we do not check that, we can just make it a required parameter. I think if we make resource id a parameter we should check if any of `$resource_( name|group|namespace)^` are valid rather than just checking 2 possible options. Again I am going off the old message so if resource_name is required we should just make it required but if one of any of those are required we can check that. Also if you want you can remove the default strings and check if they are `nil` instead. You can also do a `proc` to to the `.to_s` so you don't have to do that here as well every time you use it.

thomaslitton replied:
The resource id is made up of resource name/group/type/namespace and subscription id (along an optional resource parent).  Therefore, if you specify the resource id, the rest isn't used.  I'll make changes to make this clearer.

majormoses commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -62,7 +62,14 @@ class CheckAzurermMonitorMetric < Sensu::Plugin::Check::CLI
+         proc: proc(&:to_i),
+         default: 50_342
+
+  option :assigned_identity_resource,
```
other than the trailing `/` this looks the same as `base_url` would they ever be different? If not lets use one option and where we need the slash it can be part of the implementation itself. Maybe this can be used against an internal AD or something but I don't know much about AD and Azure to be honest.

thomaslitton replied:
I don't know if they will always be the same.  Plus one is used all the time, and one is used some of the time, so i thought it would be clearer to be 2 separate options.

majormoses replied:
ok just wanted to see if there was a valid reason and if so I am :+1:

thomaslitton committed:
Re-arranged how the resource options are validated.

Moved the logic to the front of the process.
Validated all the options in one place.
Made the message a bit clearer.

thomaslitton committed:
Report all critical/warning metrics.

tombuildsstuff commented on README.md:
````diff
@@ -428,6 +430,24 @@
+**check-azurerm-monitor-metric.rb**
+```
+./check-azurerm-monitor-metric.rb --use-assigned-identity 
+                                  --resource-id "id"
````
would it be worth making this example more complete via an example such as `/subscriptions/0000../resourceGroups/foo/...`?

tombuildsstuff commented on bin/check-azurerm-monitor-metric.rb:
```diff
@@ -0,0 +1,356 @@
+                    )
+
+                    provider.get_authentication_header
+                  end
```
not needed for this PR - but longer term we could probably split this out into a helper function?

tombuildsstuff reviewed (approved):
hey @thomaslitton

Thanks for this PR - I've taken a look through and I've left one minor comment (in addition to @majormoses comments) - but this otherwise LGTM 👍

Thanks!

thomaslitton commented:
@majormoses  Are there any more changes that are needed for this?

majormoses commented:
Sorry this fell off my radar, looke like there are some questions in the review comments still unaddressed can you address them?

majormoses reviewed (approved):
LGTM

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-azurerm/versions/2.1.0
